### PR TITLE
CASSCHEM ACTIVATE

### PR DESCRIPTION
**Pull request by KayeArray**
:cl: banthebantz
add: Ported modified lcasschem from oldcode, misc science is now a chem lab
/:cl:

casschem

### CODE REVIEW

**Comment by KayeArray** [comment]
thank fuck it's a miracle

File: code/modules/reagents/chemistry/holder.dm [hunk 1]
```diff
@@ -321,7 +321,7 @@
 		var/datum/reagent/R = reagent
 		R.on_update (A)
 	update_total()
-
+/*
```
Comment (ghost) [inline code/modules/reagents/chemistry/holder.dm]: Please add a comment why exactly this is commented out if you can

File: hippiestation/code/modules/reagents/chemistry/holder.dm [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+				var/datum/chemical_reaction/C = reaction
+				var/list/cached_required_reagents = C.required_reagents
+				var/total_required_reagents = cached_required_reagents.len
+				var/total_matching_reagents = 0
+				var/list/cached_required_catalysts = C.required_catalysts
+				var/total_required_catalysts = cached_required_catalysts.len
+				var/total_matching_catalysts= 0
+				var/matching_container = 0
+				var/matching_other = 0
+				var/list/multipliers = new/list()
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/holder.dm:42]: Please use the LAZY list helpers for creating and working with lists -- https://github.com/tgstation/tgstation/blob/master/code/__HELPERS/_lists.dm#L470

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+	if(beaker)
+		if(beaker.reagents.chem_pressure != pressure)
+			beaker.reagents.chem_pressure = pressure
+			beaker.reagents.chem_pressure = round(beaker.reagents.chem_pressure)
+			beaker.reagents.handle_reactions()
+	if(!on)
+		pressure = max(pressure -= 5, 0)
+
+/obj/machinery/chem/pressure/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, \
+										datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: Please do not break this in two lines, keep it on one line please

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+	data["isActive"] = on
+	data["isBeakerLoaded"] = beaker ? 1 : 0
+	data["internalPressure"] = pressure
+	data["currentPressure"] = beaker ? beaker.reagents.chem_pressure : null
+	data["beakerCurrentVolume"] = beaker ? beaker.reagents.total_volume : null
+	data["beakerMaxVolume"] = beaker ? beaker.volume : null
+
+	var beakerContents[0]
+	if(beaker)
+		for(var/datum/reagent/R in beaker.reagents.reagent_list)
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: Please turn this into a typeless loop

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+			return
+		to_chat(user, "<span class='notice'>You add the uranium to the [src].</span>")
+		var/obj/item/stack/sheet/mineral/uranium/U = I
+		material_amt = Clamp(material_amt += U.amount * 1000, 0, 50000)//50 sheets max
+		user.drop_item()
+		qdel(I)//it's a var now
+		return
+
+/obj/machinery/chem/radioactive/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, \
+										datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: Please do not break these lines up

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+	data["isActive"] = on
+	data["isBeakerLoaded"] = beaker ? 1 : 0
+	data["materialAmount"] = material_amt
+	data["currentRadioactivity"] = beaker ? beaker.reagents.chem_radioactivity : null
+	data["beakerCurrentVolume"] = beaker ? beaker.reagents.total_volume : null
+	data["beakerMaxVolume"] = beaker ? beaker.volume : null
+
+	var beakerContents[0]
+	if(beaker)
+		for(var/datum/reagent/R in beaker.reagents.reagent_list)
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: istypeless

**Review by ghost — CHANGES_REQUESTED** [review]
The code is a bit iffy, I'll go over the rest in a bit

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/obj/machinery/chem/radioactive//break up in action dust I walk my brow and I strut my
+	name = "Radioactive molecular reassembler"
+	desc = "A mystical machine that changes molecules directly on the level of bonding."
+	icon_state = "radio"
+	var/material_amt = 0 //requires uranium in order to function
+	var/target_radioactivity = 0
+
+/obj/machinery/chem/radioactive/New()
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: Initialize

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+	..()
+	var/obj/item/circuitboard/machine/B = new /obj/item/circuitboard/machine/radioactive(null)
+	B.apply_default_parts(src)
+
+/obj/machinery/chem/radioactive/process()
```
Comment (MoreRobustThanYou) [inline hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: shouldn't you add this to SSobj.processing?

File: hippiestation/code/modules/reagents/chemistry/holder.dm (hunk 2 above)
Comment (KayeArray) [reply hippiestation/code/modules/reagents/chemistry/holder.dm:42]: this is part of the original tg holder proc

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/obj/machinery/chem/bluespace
+	name = "Bluespace recombobulator"
+	desc = "Forget changing molecules , this thing changes the laws of physics itself in order to produce chemicals."
+	icon_state = "blue"
+	var/crystal_amt = 0
+	var/intensity = 0
+
+/obj/machinery/chem/bluespace/New()
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: initialize

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+/obj/machinery/chem/centrifuge
+	name = "Centrifuge"
+	desc = "Spins chemicals at high speeds to seperate them"
+	icon_state = "cent_off"
+	var/time_required = 30
+	var/time = 0
+
+/obj/machinery/chem/centrifuge/New()
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: too many New()s, always use initialize if possible

**Review by ghost — DISMISSED** [review]

File: code/modules/reagents/chemistry/holder.dm (hunk 1 above)
Comment (KayeArray) [reply code/modules/reagents/chemistry/holder.dm]: it's the tg proc that i needed to modify directly to get the custom reaction conditions to work, i copied it over to the hippie folder

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm (hunk 8 above)
Comment (banthebantz) [reply hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: doesn't it already default to that?

Comment (MoreRobustThanYou) [reply hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: tbh i dunno

File: code/modules/reagents/chemistry/holder.dm (hunk 1 above)
Comment (ghost) [reply code/modules/reagents/chemistry/holder.dm]: good, now place that in the code so someone that looks at that in the future knows whats up

File: hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm (hunk 4 above)
Comment (KayeArray) [reply hippiestation/code/modules/reagents/chemistry/machinery/lcasschem.dm]: do you mean iterate for var/X and assign it to datum/reagent in the loop?

**Review by ghost — APPROVED** [review]
I don't see any glaring errors, and you fixed the things me and Chnkr pointed out...

**Comment by KayeArray** [comment]
erik probably has more things to fix

File: hippiestation/code/modules/reagents/chemistry/recipes/toxins.dm [hunk 11]
```diff
@@ -2,4 +2,124 @@
...
+	pressure_required = 26
+
+/datum/chemical_reaction/sarin_a
+	name = "Translucent mixture"
+	id = "sarina"
+	results = list("sarina" = 3)
+	required_reagents = list("isopropyl" = 3, "methphos" = 2)
+
+/datum/chemical_reaction/sarin_b
+	name = "Extremely dilute sarin"
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/recipes/toxins.dm]: This should be `diluted sarin`

File: hippiestation/code/modules/reagents/chemistry/reagents/toxin_reagents.dm [hunk 12]
```diff
@@ -64,3 +64,303 @@
...
+	M.adjust_fire_stacks(1)
+	M.dizziness++
+	..()
+
+/datum/reagent/toxin/acid/hydrazine/on_tick()
+	data++
+	if(prob(2) && data > 40) //randomly creates small explosions or fireballs but has a delay so it doesn't just kill people while they're still mixing
+		var/location = get_turf(holder.my_atom)
+		holder.remove_reagent(src.id,5,safety = 1)
+		switch(pick("explosion", "fire"))
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/reagents/toxin_reagents.dm]: the pick can be inside the switch, and just make it pick numbers.

File: hippiestation/code/modules/reagents/chemistry/holder.dm [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+/datum/reagents/proc/handle_reactions()
+	var/list/cached_reagents = reagent_list
+	var/list/cached_reactions = GLOB.chemical_reactions_list
+	var/datum/cached_my_atom = my_atom
+	if(flags & REAGENT_NOREACT)
+		return //Yup, no reactions here. No siree.
+
+	var/reaction_occurred = 0
+	do
```
Comment (ghost) [inline hippiestation/code/modules/reagents/chemistry/holder.dm:25]: this loop is pretty fucking nasty to be fair

**Review by ghost — DISMISSED** [review]
Just some minor things, someone should go over the chems and make sure they are not too OP for research

File: hippiestation/code/modules/reagents/chemistry/holder.dm (hunk 13 above)
Comment (KayeArray) [reply hippiestation/code/modules/reagents/chemistry/holder.dm:25]: it fucking is i know

Comment (ghost) [reply hippiestation/code/modules/reagents/chemistry/holder.dm:25]: maybe make it better? use typecaches and all those things

Comment (ghost) [reply hippiestation/code/modules/reagents/chemistry/holder.dm:25]: done

File: hippiestation/code/modules/reagents/chemistry/recipes/toxins.dm (hunk 11 above)
Comment (KayeArray) [reply hippiestation/code/modules/reagents/chemistry/recipes/toxins.dm]: the name or the type?

Comment (KayeArray) [reply hippiestation/code/modules/reagents/chemistry/recipes/toxins.dm]: oh wait nvm

Comment (ghost) [reply hippiestation/code/modules/reagents/chemistry/recipes/toxins.dm]: either way i changed the name to "diluted sarin"

**Comment by ghost** [comment]
so, this re-adds it back into science and only the science chem machine does these reagents?

**Comment by KayeArray** [comment]
no, there are no unique elements only compounds, science gets the high risk machines, chemistry gets the two low risk machines

also science gets one chem dispenser only

**Comment by ghost** [comment]
alright, good to run by me +1

**Comment by Pyko1** [comment]
also why arent there pics of the map edit?

**Comment by KayeArray** [comment]
eh because jambread test merged it pretty much instantly so i didn't bother

**Comment by Pyko1** [comment]
welp the mapping might or might not need fixing, as its not have been really checked at all

**Comment by doobiedoo23** [comment]
yeah okay whatever

**Pull request merged**